src/pytorch_lightning/callbacks/model_checkpoint.py

```diff
@@ -262,13 +262,6 @@ def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: s
         if trainer.is_global_zero and stage == "fit":
             self.__warn_if_dir_not_empty(self.dirpath)
 
-        # NOTE: setting these attributes needs to happen as early as possible BEFORE reloading callback states,
-        # because the attributes are part of the state_key which needs to be fully defined before reloading.
```

Comment:
What about this comment? This requirement no longer holds if I understand correctly, because we evaluate the condition `self._should_save_on_train_epoch_end(trainer)` on the fly and hence later than requested by this note.

This is the reason why the existing tests had to be changed, right?
Can you describe the consequences this has on existing workflows?

Reply:
Good point. Yeah, this blocks making this attribute dynamic, as the callback state depends on its value.

I don't think we can proceed with this proposal

Reply:
> Can you describe the consequences this has on existing workflows?

for existing workflows, it will have a boolean value, so the dynamic behavior won't do anything.
```py
if self._save_on_train_epoch_end is not None:
    return self._save_on_train_epoch_end
```

Reply:
I had a discussion with Rohit in a call.

This is okay. There's a breaking change in that existing checkpoints might not get loaded into the callback instance because the instance will have `None` but the checkpoint data `bool`. I don't think we can do anything about it since there's no way for us to "upgrade" the checkpoint. We cannot differentiate between an user set value and an automatically set value from just the checkpoint.

An alternative would be to remove `save_on_train_epoch_end` from the state key, but that would also be a breaking change